File: solutions/zig/largest-series-product/1/largest_series_product.zig
pub const SeriesError = error{
    InvalidCharacter,
    NegativeSpan,
    InsufficientDigits,
};

pub fn largestProduct(digits: []const u8, span: i32) SeriesError!u64 {
    if (span < 0) return SeriesError.NegativeSpan;
    if (span > digits.len) return SeriesError.InsufficientDigits;
    if (span == 0) return 1;

    const span_: usize = @intCast(span);
    var max: u64 = 0;
    for (0..digits.len - span_ + 1) |i| {
        var product: u64 = 1;
        for (0..span_) |j| {
            const digit: u8 = digits[i + j];
            if (digit < '0' or digit > '9') return SeriesError.InvalidCharacter;
            product = product * (digit - '0');
        }
        if (product > max) {
            max = product;
        }
    }
    return max;
}

File: solutions/clojure/largest-series-product/1/src/largest_series_product.clj
(ns largest-series-product)

(defn largest-product [len digits]
  (cond (or (< (count digits) len) (< len 0)) (throw (IllegalArgumentException. "rejects span longer than string length"))
        (< len 0) (throw (IllegalArgumentException. "rejects negative length"))
        (and ((complement zero?) len) (empty? digits)) (throw (IllegalArgumentException. "rejects empty string and non zero span"))
        (re-find #"[^0-9]" digits) (throw (IllegalArgumentException. "rejects invalid character in digits"))
        (zero? len) 1
        :else (->> digits
                   (partition len 1)
                   (map #(apply * (map (fn [digit] (Integer/parseInt (str digit))) %)))
                   (apply max))))
